k2/csrc/cuda/array.h

```diff
@@ -95,24 +92,24 @@ class Array1 {
      time if it's a CUDA array, so use this operator sparingly.  If you know
      this is a CPU array, it would have much less overhead to index the Data()
      pointer. */
-  T operator [] (int32_t i);
+  T operator[](int32_t i);
 
-  Array1 operator [](const Array1<int32_t> &indexes) {
-    Context c = Context();
-    assert(c.IsCompatible(indexes.GetContext()));
+  Array1 operator[](const Array1<int32_t> &indexes) {
+    ContextPtr c = Context();
+    assert(c->IsCompatible(indexes.GetContext()));
     int32_t ret_dim = indexes.Dim();
     Array1<T> ans(c, ret_dim);
     const T *this_data = Data();
     T *ans_data = ans.Data();
     int32_t *indexes_data = indexes.Data();
-    auto lambda_copy_elems = [=] __host__ __device__ (int32_t i) -> void {
-       ans_data[i] = this_data[indexes_data[i]];
+    auto lambda_copy_elems = [=] __host__ __device__(int32_t i) -> void {
```

Comment:
It would be better if you specify explicitly what to be captured in the lambda.

Reply:
Sure, we will definitely do this when implementing other methods in `array.h`, the code from @danpovey now in `array.h` is just to delcare the interface

Reply:
The signature of this function says it returns `Array1`, but this function returns nothing.

Reply:
ah, it should return `ans`